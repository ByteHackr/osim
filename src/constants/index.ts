export enum CommentType {
    Public,
    Private,
    Internal,
    System,
}

export const SYSTEM_EMAIL = 'bugzilla@redhat.com';
