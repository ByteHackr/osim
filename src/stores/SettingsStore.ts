import { defineStore } from 'pinia';
import { z } from 'zod';
import { useLocalStorage } from '@vueuse/core';

import { useToastStore } from './ToastStore';

export const SettingsSchema = z.object({
  // bugzillaApiKey: z.string().length(
  //   32, {message: 'Bugzilla API Key is the wrong length!'}
  // ).optional(),
  bugzillaApiKey: z.string().optional().or(z.literal('')).default(''),
  jiraApiKey: z.string().optional().or(z.literal('')).default(''),
  showNotifications: z.boolean(),
  affectsPerPage: z.number(),
  trackersPerPage: z.number(),
  isHidingLabels: z.boolean().optional().default(false),
  privacyNoticeShown: z.boolean().default(false),
  singleCommentsView: z.boolean().default(false),
});

export type SettingsType = z.infer<typeof SettingsSchema>;

const defaultValues: SettingsType = {
  bugzillaApiKey: '',
  jiraApiKey: '',
  showNotifications: false,
  affectsPerPage: 10,
  trackersPerPage: 10,
  isHidingLabels: false,
  privacyNoticeShown: false,
  singleCommentsView: false,
};

export const useSettingsStore = defineStore('SettingsStore', () => {
  const { addToast } = useToastStore();

  const settings = useLocalStorage('OSIM::USER-SETTINGS', structuredClone(defaultValues));

  const validatedSettings = SettingsSchema.safeParse(settings.value);
  if (validatedSettings.success) {
    if (JSON.stringify(validatedSettings.data) !== JSON.stringify(settings.value)) {
      settings.value = validatedSettings.data;
    }
  } else {
    settings.value = structuredClone(defaultValues);
  }

  if (!settings.value.privacyNoticeShown) {
    addToast({
      title: 'Privacy Notice',
      body: 'OSIM transmits input information externally to Bugzilla for the purpose of retrieving bug data.' +
      ' In some cases that information may be publicly visible.',
    });
    settings.value.privacyNoticeShown = true;
  }

  function $reset() {
    settings.value = structuredClone(defaultValues);
  }

  return {
    $reset,
    settings,
  };
});
