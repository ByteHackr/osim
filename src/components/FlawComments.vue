<script setup lang="ts">
import { computed, nextTick, onMounted, ref, watch } from 'vue';
import { isDefined, watchDebounced } from '@vueuse/core';
import type { ZodJiraUserAssignableType } from '@/types/zodJira';
import LabelTextarea from '@/components/widgets/LabelTextarea.vue';
import sanitizeHtml from 'sanitize-html';
import { osimRuntime } from '@/stores/osimRuntime';
import { useUserStore } from '@/stores/UserStore';
import { DateTime } from 'luxon';
import Tabs from '@/components/widgets/Tabs.vue';
import DropDown from '@/components/widgets/DropDown.vue';
import { createCatchHandler } from '@/composables/service-helpers';
import { searchJiraUsers, jiraTaskUrl } from '@/services/JiraService';
import { type ZodFlawCommentSchemaType } from '@/types/zodFlaw';
import JiraUser from './widgets/JiraUser.vue';
import { type ZodFlawCommentType } from '@/types/zodFlaw';

const userStore = useUserStore();

const props = defineProps<{
  publicComments: ZodFlawCommentType[];
  privateComments: ZodFlawCommentType[];
  internalComments: ZodFlawCommentType[];
  internalCommentsAvailable: boolean;
  isLoadingInternalComments: boolean;
  systemComments: ZodFlawCommentType[];
  taskKey: string;
  bugzillaLink: string;
  isSaving: boolean;
}>();

enum CommentType {
    Public,
    Private,
    Internal,
    System,
}

const newComment = ref('');
const isAddingNewComment = ref(false);

const emit = defineEmits<{
  'comment:addFlawComment': [comment: string, creator: string, type: string];
  'loadInternalComments': [];
}>();

const commentLabels = computed(() => {
  return Object.keys(CommentType)
    .filter((key) => isNaN(Number(key)))
    .map((key) => `${key} Comments`);
});

const commentTooltips: Record<CommentType, string> = {
  [CommentType.Public]: 'Bugzilla Public - This comments are visible to everyone.',
  [CommentType.Private]: 'Bugzilla Private - This comments are visible to Red Hat associates.',
  [CommentType.Internal]: 'Jira Internal - This comments are visible to team members with required permissions.',
  [CommentType.System]: 'Bugzilla System - This are auto-generated private comments.',
};

const selectedTab = ref(CommentType.Public);
const handleTabChange = (index: number) => {
  selectedTab.value = index;
};

const displayedComments = computed(() => {
  switch (selectedTab.value) {
  case CommentType.Public:
    return props.publicComments;
  case CommentType.Private:
    return props.privateComments;
  case CommentType.Internal:
    return props.internalComments;
  case CommentType.System:
    return props.systemComments;
  default:
    return [];
  }
});

onMounted(async () => {
  emit('loadInternalComments');
});

async function handleCommentSave() {
  emit(
    'comment:addFlawComment',
    newComment.value,
    userStore.userName,
    CommentType[selectedTab.value],
  );
  newComment.value = '';
  isAddingNewComment.value = false;
}

function sanitize(text: string) {
  const bugzillaLink = `${osimRuntime.value.backends.bugzilla}/show_bug.cgi?id=`;

  const urlRegex = /\b(https?:\/\/[\S]+)\b/g;
  const jiraTagRegex = /\[~([^[\]]+)\]/g;
  const bugzillaRegex = /\[bug (\d+)\]/g;
  return (
    sanitizeHtml(text)
      .replace(urlRegex, '<a target="_blank" href="$1">$1</a>')
      .replace(jiraTagRegex, (match, p1) => `${p1}`)
      .replace(bugzillaRegex, `<a target="_blank" href="${bugzillaLink}$1">[bug $1]</a>`)
  );
}

const isInternalComment = computed(() => selectedTab.value === CommentType.Internal);
const refTextArea = ref<InstanceType<typeof LabelTextarea> | null>(null);
const suggestions = ref<ZodJiraUserAssignableType[]>([]);
const isLoadingSuggestions = ref(false);
const caretPos = ref(['0px', '0px']);

watchDebounced(newComment, async () => {
  if (!isDefined(newComment)
    || newComment.value === ''
    || !refTextArea.value?.elTextArea?.value
    || !isInternalComment.value
    || !props.taskKey) {
    return;
  }

  const lastWord = getLastWord();
  if (!lastWord?.startsWith('@') || lastWord.length < 2) {
    return;
  }
  isLoadingSuggestions.value = true;
  const users = await searchJiraUsers(lastWord.slice(1), props.taskKey)
    .catch(createCatchHandler('Failed to load jira users', false))
    .finally(() => {
      isLoadingSuggestions.value = false;
      nextTick(() => refTextArea.value?.elTextArea?.focus());
    });

  if (!users?.data?.users) {
    return;
  }
  calculateDropdownPosition(lastWord);
  nextTick(() => {
    suggestions.value = users.data?.users;
  });

}, { debounce: 500 });

const calculateDropdownPosition = (lastWord: string) => {
  const elTextArea = refTextArea.value!.elTextArea;
  const selectionStart = elTextArea?.selectionStart ?? 0;

  const textAreaRect = elTextArea!.getBoundingClientRect();
  const textAreaStyle = getComputedStyle(elTextArea!);
  const paddingLeft = Number.parseInt(textAreaStyle.getPropertyValue('padding-left').replace('px', ''));
  const paddingTop = Number.parseInt(textAreaStyle.getPropertyValue('padding-top').replace('px', ''));
  const lineHeight = Number.parseInt(textAreaStyle.getPropertyValue('line-height').replace('px', ''));

  const textAreaOffset = (textAreaRect.top ?? 0) - (elTextArea!.parentElement?.getBoundingClientRect()?.top ?? 0);
  const horizontalPos = selectionStart - newComment.value.slice(0, selectionStart).lastIndexOf('\n') - lastWord.length;
  const verticalPos = elTextArea!.value.slice(0, selectionStart).split('\n').length ?? 0;

  caretPos.value[0] = `calc(${paddingLeft}px + ${horizontalPos}ch)`;
  caretPos.value[1] = (textAreaOffset + paddingTop + Math.min(lineHeight * verticalPos, textAreaRect.height)) + 'px';
};

const handleSuggestionClick = (user: any) => {
  const lastWord = getLastWord();
  if (!lastWord) {
    return;
  }

  const username = `[~${user.name}]`;
  const contentBefore = newComment.value.slice(0, newComment.value.lastIndexOf(lastWord));
  const contentAfter = newComment.value.slice(newComment.value.lastIndexOf(lastWord) + lastWord.length);

  newComment.value = contentBefore + username + contentAfter;

  suggestions.value = [];
  refTextArea.value?.elTextArea?.focus();
};

const getLastWord = () => {
  const elTextArea = refTextArea.value?.elTextArea;
  const selectionStart = elTextArea?.selectionStart ?? 0;

  return newComment.value?.slice(0, selectionStart).split('\n')?.pop()?.split(' ')?.pop();
};

const clearSuggestions = (event: FocusEvent | KeyboardEvent | MouseEvent) => {
  if (event instanceof KeyboardEvent
    && ['Escape', 'ArrowUp', 'ArrowDown', 'ArrowLeft', 'ArrowRight', 'Enter', 'Tab']
      .includes(event.key)) {
    suggestions.value = [];
    return;
  }

  if (event instanceof MouseEvent) {
    suggestions.value = [];
    return;
  }

  if (event instanceof FocusEvent) {
    if (
      event.relatedTarget !== event.currentTarget
      && !(event.currentTarget as Node)?.contains(event.relatedTarget as Node)
    ) {
      suggestions.value = [];
    }
  }
};
</script>

<template>
  <section class="osim-comments my-2">
    <h4 class="mb-4">Comments</h4>
    <Tabs
      :labels="commentLabels"
      :default="0"
      :tooltips="Object.values(commentTooltips)"
      @tab-change="handleTabChange"
    >
      <template #header-actions>
        <div class="tab-actions">
          <button
            v-if="(
              !isAddingNewComment
              && selectedTab !== CommentType.System)
              && (internalCommentsAvailable || selectedTab !== CommentType.Internal
              )"
            type="button"
            class="btn btn-secondary tab-btn"
            :disabled="isSaving"
            @click="isAddingNewComment = true"
          >
            Add {{ CommentType[selectedTab] }} Comment
          </button>
          <a
            v-if="(selectedTab === CommentType.Public || selectedTab === CommentType.Private)"
            :href="bugzillaLink"
            target="_blank"
            class="btn btn-secondary tab-btn"
            :disabled="isSaving"
          >
            View on Bugzilla
          </a>
          <a
            v-if="(selectedTab === CommentType.Internal && internalCommentsAvailable)"
            :href="jiraTaskUrl(taskKey)"
            target="_blank"
            class="btn btn-secondary tab-btn"
            :disabled="isSaving"
          >
            View on Jira
          </a>
        </div>
      </template>
      <template #tab-content>
        <div class="tab-content">
          <div
            v-if="(
              isAddingNewComment
              && selectedTab !== CommentType.System)
              && (internalCommentsAvailable
                || selectedTab !== CommentType.Internal
              )"
            class="position-relative"
            tabindex="-1"
            @blur.capture="clearSuggestions"
          >
            <template v-if="!isInternalComment">
              <LabelTextarea v-model="newComment" :label="`New ${CommentType[selectedTab]} Comment`" />
            </template>
            <template v-if="isInternalComment">
              <LabelTextarea
                ref="refTextArea"
                v-model="newComment"
                :label="`New ${CommentType[selectedTab]} Comment`"
                :disabled="isLoadingSuggestions"
                :loading="isLoadingSuggestions"
                @keydown="clearSuggestions"
                @click="clearSuggestions"
              />
              <DropDown
                v-if="suggestions.length > 0"
                :style="{ width: 'auto', left: caretPos[0], top: caretPos[1], right: 'auto' }"
              >
                <div
                  v-for="user in suggestions"
                  :key="user.name"
                  class="item"
                  @click="handleSuggestionClick(user)"
                >
                  <JiraUser v-bind="user" :query="getLastWord()?.slice(1)" />
                </div>
              </DropDown>
            </template>
            <button type="button" class="btn btn-primary col" @click="handleCommentSave">
              Save {{ CommentType[selectedTab] }} Comment
            </button>
            <button type="button" class="btn ms-3 btn-secondary col" @click="isAddingNewComment = false">
              Cancel
            </button>
          </div>
          <ul class="comments list-unstyled">
            <span
              v-if="isLoadingInternalComments && selectedTab === CommentType.Internal"
              class="spinner-border spinner-border-sm d-inline-block ms-3"
              role="status"
            >
              <span class="visually-hidden">Loading...</span>
            </span>
            <div v-else-if="!internalCommentsAvailable && selectedTab === CommentType.Internal" class="ms-3">
              Internal comments not available
            </div>
            <div v-else-if="displayedComments?.length === 0" class="ms-3">
              No {{ CommentType[selectedTab].toLowerCase() }} comments
            </div>
            <li
              v-for="(comment, commentIndex) in displayedComments"
              :key="commentIndex"
              class="bg-light p-4 mt-3 rounded-2"
            >
              <p class="border-bottom pb-3">
                <i class="bi bi-caret-right-fill"></i>
                {{ comment.creator }}
                - {{ DateTime.fromISO(comment.created_dt ?? '',{ setZone: true })
                  .toFormat('yyyy-MM-dd hh:mm a ZZZZ') }}
                <span
                  class="badge rounded-pill float-end cursor-pointer"
                  :class="{
                    'bg-success': selectedTab === CommentType.Public,
                    'bg-info': selectedTab === CommentType.Private,
                    'bg-danger': selectedTab === CommentType.Internal,
                    'bg-warning text-black': selectedTab === CommentType.System,
                  }"
                >
                  {{ CommentType[selectedTab] }}
                </span>
              </p>
              <!--eslint-disable-next-line vue/no-v-html -->
              <p class="osim-flaw-comment" v-html="sanitize(comment.text ?? '')" />
            </li>
          </ul>
        </div>
      </template>
    </Tabs>
  </section>
</template>

<style scoped lang="scss">
.tab-actions {
  margin-left: auto;

  .tab-btn {
    margin-left: 3px;
    margin-top: 3px;
    border-bottom-right-radius: 0;
    border-bottom-left-radius: 0;
  }
}

.tab-content {
  min-height: 250px;
  margin-top: 25px;
}

.osim-comments {
  .osim-comment-filter {
    text-transform: capitalize;
  }

  section.osim-comments {
    li {
      list-style-type: none;
      background-color: red;
    }
  }

  .osim-flaw-comment {
    white-space: pre-wrap;
  }
}
</style>
