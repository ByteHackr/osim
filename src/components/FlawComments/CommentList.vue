<script setup lang="ts">
import { DateTime } from 'luxon';
import sanitizeHtml from 'sanitize-html';

import { CommentTypeDisplay } from '@/composables/useFlawCommentsModel';

import { type ZodFlawCommentType } from '@/types/zodFlaw';
import { CommentType } from '@/types';
import { jiraUserUrl } from '@/services/JiraService';
import { commentTooltips } from '@/constants';
import { osimRuntime } from '@/stores/osimRuntime';

defineProps<{
  commentList: ZodFlawCommentType[];
}>();

const parseCommentDisplayText = (text: string) => parseJiraTags(linkify(sanitizeHtml(text)));

function linkify(text: string) {
  const bugzillaLink = `${osimRuntime.value.backends.bugzilla}/show_bug.cgi?id=`;

  const bugzillaRegex = /\[bug (\d+)\]/g;
  const jiraRegex = /\[([^|]+)\|([^\]]+)\]/g;

  return text
    .replace(bugzillaRegex, `<a target="_blank" href="${bugzillaLink}$1">[bug $1]</a>`)
    .replace(jiraRegex, '<a target="_blank" href="$2">$2</a>');
}

function parseJiraTags(text: string) {
  const jiraUserLink = `${osimRuntime.value.backends.jiraDisplay}/ViewProfile.jspa?name=`;
  const jiraTagRegex = /\[~([^[\]]+)\]/g;

  return text
    .replace(jiraTagRegex, (match, p1) => `<a target="_blank" href="${jiraUserLink}${p1}">${p1}</a>`);
}

function getCommentTooltip(type: CommentType | undefined) {
  if (!type) {
    return '';
  }

  return commentTooltips[type];
}
</script>

<template>
  <ul class="comments list-unstyled">
    <li
      v-for="(comment, commentIndex) in commentList"
      :key="commentIndex"
      class="p-4 mt-3 rounded-2"
      :class="{
        'bg-light-green': comment.type === CommentType.Public,
        'bg-light-info': comment.type === CommentType.Private,
        'bg-light-orange': comment.type === CommentType.Internal,
        'bg-light-yellow': comment.type === CommentType.System,
      }"
    >
      <div class="d-flex border-bottom pb-3 align-items-center">
        <i class="bi bi-caret-right-fill me-2"></i>
        <a
          v-if="comment.type === CommentType.Internal"
          :href="jiraUserUrl(comment.creator || '')"
          target="_blank"
        >
          {{ comment.creator }}
        </a>
        <span v-else>{{ comment.creator }}</span>
        - {{ DateTime.fromISO(comment.created_dt ?? '',{ setZone: true })
          .toFormat('yyyy-MM-dd hh:mm a ZZZZ') }}
        <span
          class="badge rounded-pill float-end ms-auto"
          style="font-size: 14px; cursor: help; user-select: none;"
          :title="getCommentTooltip(comment.type)"
          :class="{
            'bg-success': comment.type === CommentType.Public,
            'bg-info': comment.type === CommentType.Private,
            'bg-danger': comment.type === CommentType.Internal,
            'bg-warning text-black': comment.type === CommentType.System,
          }"
        >
          {{ CommentTypeDisplay(comment.type) }}
        </span>
      </div>
      <!--eslint-disable-next-line vue/no-v-html -->
      <p class="osim-flaw-comment" v-html="parseCommentDisplayText(comment.text ?? '')" />
    </li>
  </ul>
</template>
