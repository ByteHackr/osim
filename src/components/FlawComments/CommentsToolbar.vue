<script setup lang="ts">
import { storeToRefs } from 'pinia';

import { jiraTaskUrl } from '@/services/JiraService';
import { useSettingsStore } from '@/stores/SettingsStore';
import { CommentType } from '@/constants';

defineProps<{
  bugzillaLink: string;
  internalCommentsAvailable: boolean;
  isAddingNewComment: boolean;
  isSaving: boolean;
  newCommentAllowed: boolean;
  newCommentType: CommentType;
  showBugzillaLink: boolean;
  showJiraLink: boolean;
  taskKey: null | string | undefined;
}>();

const emit = defineEmits<{
  startNewComment: [type: CommentType];
}>();

const { settings } = storeToRefs(useSettingsStore());
</script>

<template>
  <div class="d-flex mb-4 sticky-top p-3 ps-0 border-bottom align-items-center" style="background-color: white;">
    <div class="d-flex gap-2">
      <button
        type="button"
        class="btn btn-secondary"
        :disabled="!newCommentAllowed || isSaving || (isAddingNewComment && newCommentType === CommentType.Public)"
        @click="emit('startNewComment', CommentType.Public)"
      >
        Add Public Comment
      </button>
      <button
        type="button"
        class="btn btn-secondary"
        :disabled="!newCommentAllowed || isSaving || (isAddingNewComment && newCommentType === CommentType.Private)"
        @click="emit('startNewComment', CommentType.Private)"
      >
        Add Private Comment
      </button>
      <button
        type="button"
        class="btn btn-secondary"
        :disabled="
          !internalCommentsAvailable
            || !newCommentAllowed
            || isSaving
            || (isAddingNewComment && newCommentType === CommentType.Internal)
        "
        @click="emit('startNewComment', CommentType.Internal)"
      >
        Add Internal Comment
      </button>
      <a
        v-if="showBugzillaLink"
        :href="bugzillaLink"
        target="_blank"
        class="btn tab-btn border border-secondary"
        :disabled="isSaving"
      >
        View in Bugzilla
      </a>
      <a
        v-if="showJiraLink"
        :href="taskKey ? jiraTaskUrl(taskKey) : '#'"
        target="_blank"
        class="btn tab-btn border border-secondary"
        :disabled="isSaving"
      >
        View in Jira
      </a>
    </div>
    <div class="form-check form-switch ms-auto" style="font-size: 18px">
      <input
        id="singleViewSwitch"
        v-model="settings['unifiedCommentsView']"
        class="form-check-input"
        type="checkbox"
      >
      <label class="form-check-label" for="singleViewSwitch" style="user-select: none;">Unified view</label>
    </div>
  </div>
</template>
