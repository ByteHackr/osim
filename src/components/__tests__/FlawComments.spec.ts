import { beforeEach, describe, it, expect, vi } from 'vitest';
import { flushPromises, mount } from '@vue/test-utils';

import FlawComments from '@/components/FlawComments/FlawComments.vue';

import { type ZodFlawCommentType } from '@/types/zodFlaw';
import { searchJiraUsers } from '@/services/JiraService';
import { useSettingsStore } from '@/stores/SettingsStore';
import { mountWithConfig } from '@/__tests__/helpers';
import { CommentType } from '@/constants';

vi.mock('@/services/JiraService', () => ({
  searchJiraUsers: vi.fn(() => Promise.resolve([])),
  jiraTaskUrl: vi.fn((taskKey: string) => `http://jira-backend/browse/${taskKey}`),
  jiraUserUrl: vi.fn(),
}));

describe('flawComments', () => {
  let subject: any;
  const SYSTEM_EMAIL = 'bugzilla@redhat.com';

  beforeEach(() => {
    vi.useFakeTimers();
    subject = mountWithConfig(FlawComments, {
      props: {
        commentsByType: {
          [CommentType.Public]: [],
          [CommentType.Private]: [],
          [CommentType.Internal]: [],
          [CommentType.System]: [],
        } as Record<CommentType, ZodFlawCommentType[]>,
        internalCommentsAvailable: true,
        isLoadingInternalComments: false,
        taskKey: 'sampleKey',
        bugzillaLink: 'sampleBzLink',
        isSaving: false,
      },
    });
  });

  afterEach(() => {
    vi.resetAllMocks();
    vi.useRealTimers();
  });

  it('shows 3 correct tabs on comments section', () => {
    const navLinks = subject.findAll('.nav-link');
    expect(navLinks.length).toBe(4);
    expect(navLinks[0].text()).toBe('Public Comments');
    expect(navLinks[1].text()).toBe('Private Comments');
    expect(navLinks[2].text()).toBe('Internal Comments');
    expect(navLinks[3].text()).toBe('System Comments');
  });

  it('tab navigation changes correctly', async () => {
    const navLinks = subject.findAll('.nav-link');
    // Public comments tab (default)
    let activeTab = subject.find('.nav-link.active');
    expect(activeTab.text()).toContain('Public');
    // Private comments tab
    await navLinks[1].trigger('click');
    activeTab = subject.find('.nav-link.active');
    expect(activeTab.text()).toContain('Private');
    // Internal comments tab
    await navLinks[2].trigger('click');
    activeTab = subject.find('.nav-link.active');
    expect(activeTab.text()).toContain('Internal');
    // System comments tab
    await navLinks[3].trigger('click');
    activeTab = subject.find('.nav-link.active');
    expect(activeTab.text()).toContain('System');
  });

  it('show a functional add comment button', async () => {
    const addCommentButton = subject.find('button');
    expect(addCommentButton.exists()).toBeTruthy();
    expect(addCommentButton.text()).toBe('Add Public Comment');
    // Check that it is functional
    await addCommentButton.trigger('click');
    const newCommentInput = subject.find('textarea');
    expect(newCommentInput.exists()).toBeTruthy();
  });

  it('show message if no comments', () => {
    const commentElements = subject.findAll('ul.comments li');
    expect(commentElements.length).toBe(0);
    const noCommentsMessage = subject.find('.info-message div');
    expect(noCommentsMessage.text()).toBe('No public comments');
  });

  it('correctly display public comments', () => {
    subject = mount(FlawComments, {
      props: {
        commentsByType: {
          [CommentType.Public]: [
            { creator: 'noonerelevant', text: 'First public comment', created_dt: '2023-09-20T14:50:50Z' },
            { creator: 'onelessrelevant', text: 'Second public comment', created_dt: '2021-07-29T14:50:50Z' },
          ] as ZodFlawCommentType[],
          [CommentType.Private]: [],
          [CommentType.Internal]: [],
          [CommentType.System]: [],
        },
        internalCommentsAvailable: false,
        isLoadingInternalComments: false,
        taskKey: 'sampleKey',
        bugzillaLink: 'sampleBzLink',
        isSaving: false,
      },
    });

    const commentElements = subject.findAll('ul.comments li');
    expect(commentElements.length).toBe(2);
    // First public comment checks
    const firstHeader = commentElements[0].find('div');
    const firstBody = commentElements[0].find('p');
    expect(firstHeader.text()).toBe('noonerelevant - 2023-09-20 02:50 PM UTC');
    expect(firstBody.text()).toBe('First public comment');
    // Second public comment checks
    const secondHeader = commentElements[1].findAll('div')[0];
    const secondBody = commentElements[1].findAll('p')[0];
    expect(secondHeader.text()).toBe('onelessrelevant - 2021-07-29 02:50 PM UTC');
    expect(secondBody.text()).toBe('Second public comment');
  });

  it('correctly display private comments', async () => {
    subject = mount(FlawComments, {
      props: {
        commentsByType: {
          [CommentType.Public]: [],
          [CommentType.Private]: [
            { creator: 'noonerelevant', text: 'First private comment', created_dt: '2021-07-29T15:50:50Z' },
            { creator: 'onelessrelevant', text: 'Second private comment', created_dt: '2023-09-20T15:50:50Z' },
          ] as ZodFlawCommentType[],
          [CommentType.Internal]: [],
          [CommentType.System]: [],
        },
        internalCommentsAvailable: false,
        isLoadingInternalComments: false,
        taskKey: 'sampleKey',
        bugzillaLink: 'sampleBzLink',
        isSaving: false,
      },
    });

    const navLinks = subject.findAll('.nav-link');
    await navLinks[1].trigger('click');
    expect(subject.html()).toMatchSnapshot();
  });

  it('show Jira link button', async () => {
    subject = mount(FlawComments, {
      props: {
        commentsByType: {
          [CommentType.Public]: [],
          [CommentType.Private]: [],
          [CommentType.Internal]: [
            { creator: 'noonerelevant', text: 'First internal comment', created_dt: '2023-09-20T16:50:50Z' },
            { creator: 'onelessrelevant', text: 'Second internal comment', created_dt: '2021-07-29T16:50:50Z' },
          ] as ZodFlawCommentType[],
          [CommentType.System]: [],
        },
        internalCommentsAvailable: true,
        isLoadingInternalComments: false,
        taskKey: 'sampleKey',
        bugzillaLink: 'sampleBzLink',
        isSaving: false,
      },
    });
    const jiraLinkBtn = subject.findAll('a')[0];
    expect(jiraLinkBtn.text()).toBe('View in Jira');
    expect(jiraLinkBtn.exists()).toBeTruthy();
  });

  it('correctly display internal comments', async () => {
    subject = mount(FlawComments, {
      props: {
        commentsByType: {
          [CommentType.Public]: [],
          [CommentType.Private]: [],
          [CommentType.Internal]: [
            { creator: 'noonerelevant', text: 'First internal comment', created_dt: '2023-09-20T16:50:50Z' },
            { creator: 'onelessrelevant', text: 'Second internal comment', created_dt: '2021-07-29T16:50:50Z' },
          ] as ZodFlawCommentType[],
          [CommentType.System]: [],
        },
        internalCommentsAvailable: true,
        isLoadingInternalComments: false,
        taskKey: 'sampleKey',
        bugzillaLink: 'sampleBzLink',
        isSaving: false,
      },
    });
    const navLinks = subject.findAll('.nav-link');
    await navLinks[2].trigger('click');
    const commentElements = subject.findAll('ul.comments li');
    expect(commentElements.length).toBe(2);
    // First internal comment checks
    const firstHeader = commentElements[0].findAll('div')[0];
    const firstBody = commentElements[0].findAll('p')[0];
    expect(firstHeader.text()).toBe('noonerelevant - 2023-09-20 04:50 PM UTC');
    expect(firstBody.text()).toBe('First internal comment');
    // Seccond internal comment checks
    const secondHeader = commentElements[1].findAll('div')[0];
    const secondBody = commentElements[1].findAll('p')[0];
    expect(secondHeader.text()).toBe('onelessrelevant - 2021-07-29 04:50 PM UTC');
    expect(secondBody.text()).toBe('Second internal comment');
  });

  it('don\'Show any action button on system comments', async () => {
    const navLinks = subject.findAll('.nav-link');
    await navLinks[3].trigger('click');
    const actionButtons = subject.findAll('.tab-actions > *');
    expect(actionButtons.length).toBe(0);
  });

  it('show message if no system comments', async () => {
    const navLinks = subject.findAll('.nav-link');
    await navLinks[3].trigger('click');
    const commentElements = subject.findAll('ul.comments li');
    expect(commentElements.length).toBe(0);
    const noCommentsMessage = subject.find('.info-message div');
    expect(noCommentsMessage.text()).toBe('No system comments');
  });

  it('correctly display system comments', async () => {
    subject = mount(FlawComments, {
      props: {
        commentsByType: {
          [CommentType.Public]: [],
          [CommentType.Private]: [],
          [CommentType.Internal]: [],
          [CommentType.System]: [
            { creator: SYSTEM_EMAIL, text: 'First system comment', created_dt: '2023-09-20T17:50:50Z' },
            { creator: SYSTEM_EMAIL, text: 'Second system comment', created_dt: '2021-07-29T17:50:50Z' },
          ] as ZodFlawCommentType[],
        },

        internalCommentsAvailable: false,
        isLoadingInternalComments: false,
        taskKey: 'sampleKey',
        bugzillaLink: 'sampleBzLink',
        isSaving: false,
      },
    });

    const navLinks = subject.findAll('.nav-link');
    await navLinks[3].trigger('click');
    const commentElements = subject.findAll('ul.comments li');
    expect(commentElements.length).toBe(2);
    // First system comment checks
    const firstHeader = commentElements[0].findAll('div')[0];
    const firstBody = commentElements[0].findAll('p')[0];
    expect(firstHeader.text()).toBe(`${SYSTEM_EMAIL} - 2023-09-20 05:50 PM UTC`);
    expect(firstBody.text()).toBe('First system comment');
    // Seccond system comment checks
    const secondHeader = commentElements[1].findAll('div')[0];
    const secondBody = commentElements[1].findAll('p')[0];
    expect(secondHeader.text()).toBe(`${SYSTEM_EMAIL} - 2021-07-29 05:50 PM UTC`);
    expect(secondBody.text()).toBe('Second system comment');
  });

  it('should call `searchJiraUsers` on internal comments', async () => {
    vi.mocked(searchJiraUsers, { partial: true }).mockResolvedValueOnce({
      data: {
        users: [{ name: 'test', displayName: 'Test User', avatarUrl: '' }],
      },
    });

    const navLinks = subject.findAll('button');
    await navLinks[2].trigger('click');

    const newCommentInput = subject.find('textarea');
    await newCommentInput.setValue('test @user');

    vi.runAllTimers();
    await flushPromises();

    expect(searchJiraUsers).toHaveBeenCalledWith('user', 'sampleKey');
  });

  it('should not render tabs if unified view is set', async () => {
    const store = useSettingsStore();
    store.settings.unifiedCommentsView = true;
    await flushPromises();

    const tabs = subject.find('.nav-tabs');
    expect(tabs.exists()).toBeFalsy();
  });
});
