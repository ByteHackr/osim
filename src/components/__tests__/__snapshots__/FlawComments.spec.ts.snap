// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FlawComments > Correctly display private comments 1`] = `
"<section data-v-bb90633e="" class="osim-comments my-2">
  <h4 data-v-bb90633e="" class="mb-4">Comments</h4>
  <div data-v-8cd9b350="" data-v-bb90633e="">
    <ul data-v-8cd9b350="" class="nav nav-tabs">
      <li data-v-8cd9b350="" class="nav-item"><button data-v-8cd9b350="" type="button" class="nav-link" title="Bugzilla Public - This comments are visible to everyone.">Public Comments</button></li>
      <li data-v-8cd9b350="" class="nav-item"><button data-v-8cd9b350="" type="button" class="nav-link active" title="Bugzilla Private - This comments are visible to Red Hat associates.">Private Comments</button></li>
      <li data-v-8cd9b350="" class="nav-item"><button data-v-8cd9b350="" type="button" class="nav-link" title="Jira Internal - This comments are visible to team members with required permissions.">Internal Comments</button></li>
      <li data-v-8cd9b350="" class="nav-item"><button data-v-8cd9b350="" type="button" class="nav-link" title="Bugzilla System - This are auto-generated private comments.">System Comments</button></li>
      <div data-v-bb90633e="" class="tab-actions"><button data-v-bb90633e="" type="button" class="btn btn-secondary tab-btn"> Add Private Comment </button><a data-v-bb90633e="" href="sampleBzLink" target="_blank" class="btn btn-secondary tab-btn" disabled="false"> View on Bugzilla </a>
        <!--v-if-->
      </div>
    </ul>
    <div data-v-bb90633e="" class="tab-content">
      <div data-v-bb90633e="" class="info-message ms-3 mb-4">
        <!--v-if-->
      </div>
      <!--v-if-->
      <ul data-v-bb90633e="" class="comments list-unstyled">
        <li data-v-bb90633e="" class="bg-light p-4 mt-3 rounded-2">
          <p data-v-bb90633e="" class="border-bottom pb-3"><i data-v-bb90633e="" class="bi bi-caret-right-fill me-2"></i><span data-v-bb90633e="">noonerelevant</span> - 2021-07-29 03:50 PM UTC <span data-v-bb90633e="" class="badge rounded-pill float-end cursor-pointer bg-info">Private</span></p>
          <!--eslint-disable-next-line vue/no-v-html -->
          <p data-v-bb90633e="" class="osim-flaw-comment">First private comment</p>
        </li>
        <li data-v-bb90633e="" class="bg-light p-4 mt-3 rounded-2">
          <p data-v-bb90633e="" class="border-bottom pb-3"><i data-v-bb90633e="" class="bi bi-caret-right-fill me-2"></i><span data-v-bb90633e="">onelessrelevant</span> - 2023-09-20 03:50 PM UTC <span data-v-bb90633e="" class="badge rounded-pill float-end cursor-pointer bg-info">Private</span></p>
          <!--eslint-disable-next-line vue/no-v-html -->
          <p data-v-bb90633e="" class="osim-flaw-comment">Second private comment</p>
        </li>
      </ul>
    </div>
  </div>
</section>"
`;
